### ISSUE TITLE
Reflect indent level using VoiceOver

### ISSUE THREAD

**Issue by ivanfetch** [microsoft/vscode#88877]
A verbal indication of the indent level while using the arrow-keys to navigate code is very valuable, especially when indentation is syntactically required such as with YAML or Python.

For example, using the [vim](https://www.vim.org/) editor, I display the indent level using the [DynamicSigns plugin](https://github.com/chrisbra/DynamicSigns.git). I also navigate to lines of lesser, equal, or
greater indent using the [IndentWise plugin](https://github.com/jeetsukumaran/vim-indentwise).

CC @isidorn

**Comment by isidorn** [comment]
@ivanfetch thanks a lot for providing this valuable feedback. Would you expect that
1) indentation is always read out as you navigate the document? This has the downside of potentially being too spammy to the user. How to determine a good time to read the indentation? 
2) There is an explicit command that reads out the indentation once a user triggers it? A user could easily add a shortcut to this command.

Let me know what you think.

As for Vim, vscode also has a vim plugin https://marketplace.visualstudio.com/items?itemName=vscodevim.vim
However I am not sure if the vscode vim plugin supports extension you are using. @octref and @rebornix would be better to answer this one (as I do not use VIM).

**Comment by ivanfetch** [comment]
> @ivanfetch thanks a lot for providing this valuable feedback. Would you expect that
> 
> 1. indentation is always read out as you navigate the document? This has the downside of potentially being too spammy to the user. How to determine a good time to read the indentation?

I think this being a toggled feature (with hotkey) provides the most flexibility. There are cases where I want to know the indent level of every line as I navigate a file, but I wouldn't want that 100% of the time.

> 2. There is an explicit command that reads out the indentation once a user triggers it? A user could easily add a shortcut to this command.

A hotkey that reads the indent level of the current line, but does not toggle the "always on" feature would be a nice addition. This would exist along with the above option (to have indent level always spoken)

> As for Vim, vscode also has a vim plugin https://marketplace.visualstudio.com/items?itemName=vscodevim.vim
> However I am not sure if the vscode vim plugin supports extension you are using. @octref and @rebornix would be better to answer this one (as I do not use VIM).

I have not tried this, but the VS Code VIM Emulator appears to use emulated plugins. I think an indent feature directly in VS Code is more strait forward.

**Comment by isidorn** [comment]
Thanks. Yeah I think that makes sense. A toggle to always read indentation. Once we have that we can look into introducing the other command to only read the current line ident.
Let's assign to next milestone and we can look into adding this feature.

**Comment by rebornix** [comment]
We don't support such plugin in the Vim extension but it makes sense to have this feature in the core and keymap extensions can benefit from it directly.

**Comment by isidorn** [comment]
This is also one of the candidates for the aria-annotations.
@joanmarie @jvesouze does Orca behave better on Linux. How do users figure out the indentation? Does it make sense to introduce an explicit command for this in VS Code or should we tackl this via Aria annotations hopefully.

**Comment by joanmarie** [comment]
@isidorn Orca has a setting on each speech page to announce indentation and justification. When it is enabled, Orca announces the whitespace ("4 spaces" or "1 tab" for example) as you arrow to that line. It does this by looking at the text exposed to it for that line.

At the present time, Orca is not announcing when the indentation is inserted, but Orca could do that too if it's getting the right events. Lemme look.

(Is this what you're asking about, or somethign else?)

**Comment by isidorn** [comment]
The first part of your answer is what I am asking about. So yes Orca can announce indentation. Great!

@leonardder how about NVDA. Does NVDA have a feature to announce indentation and does it work with VS Code?

I am just thinking about a VS Code workaound for announcing indentation. Should it be only for mac and VoiceOver or for all platforms. Should we use aria annotations etc...

**Comment by joanmarie** [comment]
If there is no VoiceOver option for announcing indentation in editors, maybe there should be??

Thoughts @cookiecrook?

**Comment by cookiecrook** [comment]
Not sure. Indentation is relevant in some languages and irrelevant in others. It’s also possible for editor app developers to do this with user prefs. For example, VoiceOver speaks extra characters in Word if the user has “show invisibles” enabled. A similar approach could be used for code indentation, or the user could set up replacements in an Activity.

**Comment by isidorn** [comment]
@cookiecrook thanks for your reply. I wanted to try this "show invisibles" setting in VoiceOver but I could not find it, googling also did not help. How do I enable this?
How would the replacments in an Activity approach work, could you clarify?

**Comment by cookiecrook** [comment]
That’s not a VO setting. Show invisibles is an app specific setting (e.g. Word and several other apps have it) that renders invisible characters visibly. For example, the white space tab characters are rendered as char icons representing Tab.

**Comment by cookiecrook** [comment]
Activities is a VO feature that allows a user to change a setting or setting in the context of an app or web site. 

For example, make a new Code activity in VoiceOver Utility, and change the setting to pronounce all punctuation. Associate that activity with vscode (so it switches automatically) or select it with the activity chooser (VO+x). 

You can also set up Speech replacements for an Activity. I haven’t tried it with white space, but if the app renders Tab characters properly to the AX API, you should be able to replace these in the speech output by replacing the Tab character with something pronounceable like the word “tab.” 

Note: I haven’t tried white space replacements, so if that doesn’t work for you in vscode, let me know and I’ll look into it further.

**Comment by MarcoZehe** [comment]
You can also do replacements like two tabs with "indentation level 2", three tabs with "indentation level 3" or such I believe, make tab-indented files read really nicely. Now if we could only have punctuation spoken only in the editor, not in the settings or other parts. :-) But VO activities don't allow for that granularity yet I think.

**Comment by isidorn** [comment]
Thanks a lot for feedback.

VS Code actually already has a setting 	
`editor.renderWhitespace`  which by default is selection so we would only render whitespace when the whitespace is selected. We should change our Screen Reader behavior such that when `editor.renderWhitespace` is set to "all" we would also read the whitespace characters as they are seen in the editor.
The character which we currently use for representing a single space is a • ([bullet](https://www.compart.com/en/unicode/U+2022)). And for a tab we use ⭢ right arrow.

Does it make sense that for having three whitespaces that we read "bullet bullet bullet", or for 4 tabs we would read "right arrow right arrow right arrow right arrow". This sounds very spammy to me.

fyi @alexdima on ideas

**Comment by jhomme** [comment]
Hi,
Screen readers usually only read indentation when it changes as you move up and down. Also, if VoiceOver is set to read all punctuation, it reads tab characters in apps other than VS Code. I have not tried VS Code on Mac yet. I'll get Insider on my Mac and see what I can find out.

**Comment by isidorn** [comment]
I heard that this has arrived with the new VoiceOver with macOS Monterey. If that is not the case do let me know so I reopen this issue.

**Comment by jhomme** [comment]
Hi,
I have an early 2015 Macbook Air. If it will update to the newest OS, I
will report back.

Jim

==========

Contribute to the collection of accessible chess resources here:
https://forms.gle/idkdxJzQLZ8odq5x8
